pkg/controller/master-controller-manager/kubeone/controller.go

```diff
@@ -48,33 +55,112 @@ import (
 	"sigs.k8s.io/controller-runtime/pkg/predicate"
 	"sigs.k8s.io/controller-runtime/pkg/reconcile"
 	"sigs.k8s.io/controller-runtime/pkg/source"
+	"sigs.k8s.io/yaml"
 )
 
 const (
 	// This controller manages kubeone external cluster.
 	ControllerName = "kubeone_controller"
 	ImportAction   = "import"
+	// Upgrade control plane only.
+	UpgradeControlPlaneAction = "upgradeCP"
+	// KubeOne Pod Name.
+	KubeOneUpgradePod    = "kubeone-upgrade"
+	KubeOneKubeconfigPod = "kubeone-kubeconfig"
 )
 
 type reconciler struct {
 	ctrlruntimeclient.Client
 	log *zap.SugaredLogger
+	kubernetesprovider.ImpersonationClient
+	secretKeySelector provider.SecretKeySelectorValueFunc
 }
 
 func Add(
 	ctx context.Context,
 	mgr manager.Manager,
 	log *zap.SugaredLogger) error {
+	defaultImpersonationClient := kubernetesprovider.NewImpersonationClient(ctrlruntime.GetConfigOrDie(), mgr.GetRESTMapper())
+
 	reconciler := &reconciler{
-		log:    log.Named(ControllerName),
-		Client: mgr.GetClient(),
+		log:                 log.Named(ControllerName),
+		Client:              mgr.GetClient(),
+		ImpersonationClient: defaultImpersonationClient.CreateImpersonatedClient,
+		secretKeySelector:   provider.SecretKeySelectorValueFuncFactory(ctx, mgr.GetClient()),
 	}
 	c, err := controller.New(ControllerName, mgr, controller.Options{Reconciler: reconciler})
 	if err != nil {
 		return err
 	}
 
-	return c.Watch(&source.Kind{Type: &kubermaticv1.ExternalCluster{}}, &handler.EnqueueRequestForObject{}, withEventFilter())
+	if err := c.Watch(&source.Kind{Type: &kubermaticv1.ExternalCluster{}}, &handler.EnqueueRequestForObject{}, withEventFilter()); err != nil {
+		return fmt.Errorf("failed to create externalcluster watcher: %w", err)
+	}
+	if err := c.Watch(&source.Kind{Type: &corev1.Secret{}},
+		enqueueExternalCluster(reconciler.Client, reconciler.log),
+		withManifestEventFilter(),
+		ByNameandNamespace(),
+	); err != nil {
+		return fmt.Errorf("failed to create kubeone manifest watcher: %w", err)
+	}
+	if err := c.Watch(&source.Kind{Type: &corev1.Pod{}},
+		&handler.EnqueueRequestForOwner{
+			IsController: true,
+			OwnerType:    &kubermaticv1.ExternalCluster{},
+		},
+		withPodEventFilter(),
+	); err != nil {
+		return fmt.Errorf("failed to create kubeone pod watcher: %w", err)
+	}
+
+	return nil
+}
+
+func enqueueExternalCluster(client ctrlruntimeclient.Client, log *zap.SugaredLogger) handler.EventHandler {
+	return handler.EnqueueRequestsFromMapFunc(func(a ctrlruntimeclient.Object) []reconcile.Request {
+		externalClusterName := strings.Split(a.GetNamespace(), "-")[1]
+		return []reconcile.Request{{NamespacedName: types.NamespacedName{Name: externalClusterName, Namespace: metav1.NamespaceAll}}}
```

Comment:
This string manipulation is extremely brittle and can break super easily. It's also not clear where this string logic is coming from exactly. Can you please add a function to the same package that generates such namespaces, and call that function here?

Since you insisted (😁) on having your prefix be only "kubeone" and not "kubeone-", a namespace like "kubeonerocks" will lead to panics in this implementation.

Reply:
I understand you point, changed prefix to "kubeone-"